airbrakes/data_handling/data_processor.py

```diff
@@ -87,6 +79,138 @@ def avg_acceleration_mag(self) -> float:
+
+        # Zero the altitude only once, during the first update:
+        if self._zeroed_altitude is None:
+            self._zeroed_altitude = self._calculate_zeroed_altitude(estPressureAlt)
+
+        self._current_altitude = self._calculate_current_altitude(estPressureAlt)
+        self._max_altitude = self._calculate_max_altitude(estPressureAlt)
+
+    def _calculate_max_altitude(self, pressure_alt: list[float]) -> float:
+        """
+        Calculates the maximum altitude (zeroed out) of the rocket based on the pressure
+        altitude during the flight.
+
+        :return: The maximum altitude of the rocket in meters.
+        """
+        zeroed_alts = np.array(pressure_alt) - self._zeroed_altitude
+        return max(*zeroed_alts, self._max_altitude)
+
+    def _calculate_zeroed_altitude(self, pressure_alt: list[float]) -> float:
+        """
+        Calculates the zeroed altitude of the rocket based on the pressure altitude during startup.
+        Takes the average of the data points.
+
+        :return: The altitude of the rocket in meters, which will be used to zero the altitude.
+        """
+        # Get the initial altitude of the rocket during startup, by averaging the altitude data points
+        return np.mean(pressure_alt)
+
+    def _calculate_current_altitude(self, alt_list: list[float]) -> float:
+        """
+        Calculates the current altitude, by zeroing out the latest altitude data point.
+
+        :param alt_list: A list of the altitude data points.
+        """
+        # There is a decent chance that the zeroed out altitude is negative, e.g. if the rocket
+        # landed at a height below from where it launched from, but that doesn't concern us.
+        return alt_list[-1] - self._zeroed_altitude
+
+    def _compute_averages(self, a_x: list[float], a_y: list[float], a_z: list[float]) -> tuple[float, float, float]:
+        """
+        Calculates the average acceleration and acceleration magnitude of the data points.
+
+        :param a_x: A list of the accelerations in the x direction.
+        :param a_y: A list of the accelerations in the y direction.
+        :param a_z: A list of the accelerations in the z direction.
+
+        :return: A tuple of the average acceleration in the x, y, and z directions.
+        """
+        # calculate the average acceleration in the x, y, and z directions
+        avg_x_accel = np.mean(a_x)
+        avg_y_accel = np.mean(a_y)
+        avg_z_accel = np.mean(a_z)
+        return avg_x_accel, avg_y_accel, avg_z_accel
+
+    def _calculate_speed(self, a_x: list[float], a_y: list[float], a_z: list[float]) -> float:
+        """
+        Calculates the speed of the rocket based on the linear acceleration.
+        Integrates the linear acceleration to get the speed.
+        """
+        # We need at least two data points to calculate the speed:
+        if len(self._data_points) < 2:
+            return 0.0
```

Comment:
I'd still like us to reconsider fetching a set amount of data packets in `get_imu_data_packets()`, so that this condition is never hit and things are a little more consistent overall.

Reply:
we can tell it, "give me at least 10 packets of data, but if you have more, give it all"